src/Features/Core/Portable/IntroduceVariable/AbstractIntroduceParameterService.cs

```diff
@@ -0,0 +1,688 @@
+
+            var semanticDocument = await SemanticDocument.CreateAsync(document, cancellationToken).ConfigureAwait(false);
+
+            var expressionType = semanticDocument.SemanticModel.GetTypeInfo(expression, cancellationToken).Type;
+            if (expressionType is null or IErrorTypeSymbol)
+            {
+                return null;
+            }
+
+            var actions = AddActions(semanticDocument, expression, cancellationToken);
+
+            if (actions.Length > 0)
+            {
+                return new CodeActionWithNestedActions(FeaturesResources.Introduce_parameter, actions, isInlinable: true);
+            }
+
+            return null;
+        }
+
+        /// <summary>
+        /// Introduces a new parameter and refactors all the call sites
+        /// </summary>
+        public async Task<Solution> IntroduceParameterAsync(SemanticDocument document, TExpressionSyntax expression,
+            bool allOccurrences, bool trampoline, bool overload, CancellationToken cancellationToken)
+        {
+            var parameterName = GetNewParameterName(document, expression, cancellationToken);
+
+            // MethodSymbol not null here since we know we're contained in something containing parameters at this point
+            var methodSymbolInfo = (IMethodSymbol)document.SemanticModel.GetRequiredEnclosingSymbol(expression.SpanStart, cancellationToken);
+            var methodCallSites = await FindCallSitesAsync(document, methodSymbolInfo, cancellationToken).ConfigureAwait(false);
+
+            if (trampoline || overload)
+            {
+                return await RewriteSolutionWithNewMethodAsync(document,
+                expression, methodSymbolInfo, allOccurrences, parameterName, methodCallSites, trampoline, cancellationToken).ConfigureAwait(false);
+            }
+
+            return await RewriteSolutionAsync(document,
+                expression, methodCallSites, allOccurrences, parameterName, cancellationToken).ConfigureAwait(false);
+        }
+
+        protected static string GetNewParameterName(SemanticDocument document, TExpressionSyntax expression, CancellationToken cancellationToken)
+        {
+            if (ShouldRemoveVariableDeclaratorContainingExpression(document, expression, out var varDecl))
+            {
+                return varDecl;
+            }
+
+            var semanticFacts = document.Document.GetRequiredLanguageService<ISemanticFactsService>();
+            return semanticFacts.GenerateNameForExpression(document.SemanticModel, expression, capitalize: false, cancellationToken);
+        }
+
+        protected static bool ShouldRemoveVariableDeclaratorContainingExpression(
+            SemanticDocument document, TExpressionSyntax expression, out string varDecl)
+        {
+            var syntaxFacts = document.Document.GetRequiredLanguageService<ISyntaxFactsService>();
+            var expressionDecl = expression.GetRequiredParent().GetRequiredParent();
+            varDecl = "";
```

Comment:
```suggestion
```

Reply:
![image](https://user-images.githubusercontent.com/40616383/112394377-daa4bd80-8cb9-11eb-9759-2c01abc3f991.png)
It says I have to set the value

Reply:
removed